client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -444,44 +347,112 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
                 }
+                return hosts;
             }
-            if (filteredHosts.isEmpty()) {
-                //last resort: if there are no good host to use, return a single dead one, the one that's closest to being retried
-                List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
-                if (sortedHosts.size() > 0) {
-                    Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
-                        @Override
-                        public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
-                            return Long.compare(o1.getValue().getDeadUntilNanos(), o2.getValue().getDeadUntilNanos());
-                        }
-                    });
-                    HttpHost deadHost = sortedHosts.get(0).getKey();
-                    logger.trace("resurrecting host [" + deadHost + "]");
-                    nextHosts = Collections.singleton(deadHost);
+        }
+
+        /*
+         * Last resort: If there are no good nodes to use, either because
+         * the selector rejected all the living nodes or because there aren't
+         * any living ones. Either way, we want to revive a single dead node
+         * that the NodeSelectors are OK with. We do this by sorting the dead
+         * nodes by their revival time and passing them through the
+         * NodeSelector so it can have its say in which nodes are ok and their
+         * ordering. If the selector is ok with any of the nodes then use just
+         * the first one in the list because we only want to revive a single
+         * node.
+         */
+        if (false == deadNodes.isEmpty()) {
+            Collections.sort(deadNodes, new Comparator<DeadNodeAndRevival>() {
+                @Override
+                public int compare(DeadNodeAndRevival lhs, DeadNodeAndRevival rhs) {
+                    return Long.compare(rhs.nanosUntilRevival, lhs.nanosUntilRevival);
                 }
-            } else {
-                List<HttpHost> rotatedHosts = new ArrayList<>(filteredHosts);
-                Collections.rotate(rotatedHosts, rotatedHosts.size() - lastHostIndex.getAndIncrement());
-                nextHosts = rotatedHosts;
+            });
+
+            List<Node> selectedDeadNodes = new ArrayList<>(deadNodes.size());
+            for (DeadNodeAndRevival n : deadNodes) {
+                selectedDeadNodes.add(n.node);
             }
-        } while(nextHosts.isEmpty());
-        return new HostTuple<>(nextHosts.iterator(), hostTuple.authCache);
+            selectedDeadNodes = nodeSelector.select(selectedDeadNodes);
+            if (false == selectedDeadNodes.isEmpty()) {
+                return singletonList(selectedDeadNodes.get(0).getHost());
+            }
+        }
+        throw new IOException("NodeSelector [" + nodeSelector + "] rejcted all nodes, "
```

Comment:
if you throw above, maybe throw inside, only in case the selector returns empty here too, that might make the code a bit more readable.